File: C#.Net/JDI/Web/Selenium/Elements/Complex/table/Rows.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Epam.JDI.Commons;
using Epam.JDI.Web.Selenium.Elements.Complex.table.interfaces;
using OpenQA.Selenium;
using static Epam.JDI.Core.ExceptionUtils;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Complex.table
{
    public class Rows : TableLine
    {
        public Rows()
        {
            HasHeader = false;
            ElementIndex = ElementIndexType.Nums;
            HeadersLocator = By.XPath(".//tr/td[1]");
            DefaultTemplate = By.XPath(".//tr[{0}]/td");
        }
        
        protected override Func<TableLine, IList<IWebElement>> GetHeadersAction => 
            row => Table.WebElement.FindElements(HeadersLocator);

        protected override IList<IWebElement> GetFirstLine()
        {
            return Table.Columns.GetLineAction(1);
        }

        public override Dictionary<string, Dictionary<string, ICell>> Get()
        {
            return Headers.ToDictionary(key => key, GetRow);
        }

        ///

        public IList<string> GetRowValue(string rowName)
        {
            return ActionWithException(() => Table.Rows.GetLineAction(rowName).Select(el => el.Text).ToList(),
                ex => $"Can't Get Row '{rowName}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetRowAsText(string rowName)
        {
            return GetRow(rowName).ToDictionary(pair => pair.Key, pair => pair.Value.Text);
        }

        public Dictionary<string, ICell> CellsToRow(ICollection<ICell> cells)
        {
            return cells.ToDictionary(
                    cell => Headers[cell.ColumnNum - 1],
                    cell => cell);
        }

        public Dictionary<string, ICell> GetRow(int rowNum)
        {
            if (Count < 0 || Table.Rows.Count < rowNum || rowNum <= 0)
                throw Exception($"Can't Get Row '{rowNum}'. [num] > ColumnsCount({Count}).");
            return ActionWithException(() =>
            {
                var colsCount = Table.Columns.Count;
                var webRow = Timer.GetResultByCondition(() => GetLineAction(rowNum), els => els.Count == colsCount);
                var result = new Dictionary<string, ICell>();
                for (var i = 0; i < colsCount; i++)
                    result.Add(Table.Columns.Headers[i], Table.Cell(webRow[i], new Column(i + 1), new Row(rowNum)));
                return result;
            }, ex => $"Can't Get Row '{rowNum}'. Reason: {ex}");
        }

        public IList<string> GetRowValue(int rowNum)
        {
            if (Count < 0 || Count < rowNum || rowNum <= 0)
                throw Exception($"Can't Get Row '{rowNum}'. [num] > ColumnsCount({Count}).");
            return ActionWithException(() => Table.Rows.GetLineAction(rowNum).Select(el => el.Text).ToList(),
                ex => $"Can't Get Row '{rowNum}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetRowAsText(int rowNum)
        {
            return GetRow(rowNum).ToDictionary(pair => pair.Key, pair => pair.Value.Text);
        }

        public Dictionary<string, ICell> GetRow(string rowName)
        {
            return ActionWithException(() =>
            {
                var colsCount = Table.Columns.Count;
                var webRowLine = Timer.GetResultByCondition(() => GetLineAction(rowName), els => els.Count == colsCount);
                var headers = Table.Columns.Headers;
                var webRow = SkipFirstColumn() ? webRowLine.ListCopy(1) : webRowLine;
                var result = new Dictionary<string, ICell>();
                for (var i = 0; i < colsCount; i++)
                    result.Add(Table.Columns.Headers[i], Table.Cell(webRow[i], new Column(name: headers[i]), new Row(name: rowName)));
                return result;
            }, ex => $"Can't Get Row '{rowName}'. Reason: {ex}");
        }

        public bool SkipFirstColumn()
        {
            return HasHeader && LineTemplate == null;
        }
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winium/elements/complex/table/Rows.java
package com.epam.jdi.uitests.win.winium.elements.complex.table;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.win.winium.elements.complex.table.interfaces.ICell;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.Collection;
import java.util.List;

import static com.epam.commons.LinqUtils.select;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by 12345 on 26.10.2014.
 */
public class Rows extends TableLine {
    public Rows() {
        hasHeader = false;
        elementIndex = ElementIndexType.Nums;
        headersLocator = By.xpath(".//tr/td[1]");
        defaultTemplate = By.xpath(".//tr[%s]/td");
    }

    protected List<WebElement> getHeadersAction() {
        return table.getWebElement().findElements(headersLocator);
    }

    protected List<WebElement> getFirstLine() {
        return table.columns().getLineAction(1);
    }

    public MapArray<String, MapArray<String, ICell>> get() {
        return new MapArray<>(headers(), key -> key, this::getRow);
    }

    ///

    public List<String> getRowValue(String rowName) {
        try {
            return select(table.rows().getLineAction(rowName), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwRowsException(rowName, ex.getMessage());
        }
    }

    public final MapArray<String, String> getRowAsText(String rowName) {
        return getRow(rowName).toMapArray(IText::getText);
    }

    public MapArray<String, ICell> cellsToRow(Collection<ICell> cells) {
        return new MapArray<>(cells,
                cell -> headers().get(cell.columnNum() - 1),
                cell -> cell);
    }

    public MapArray<String, ICell> getRow(int rowNum) {
        if (count() < 0 || table.rows().count() < rowNum || rowNum <= 0)
            throw exception("Can't Get Row '%s'. [num] > ColumnsCount(%s).", rowNum, count());
        try {
            int colsCount = table.columns().count();
            List<WebElement> webRow = timer().getResultByCondition(
                    () -> getLineAction(rowNum), els -> els.size() == colsCount);
            return new MapArray<>(colsCount,
                    key -> table.columns().headers().get(key),
                    value -> table.cell(webRow.get(value), new Column(value + 1), new Row(rowNum)));
        } catch (Exception | Error ex) {
            throw throwRowsException(Integer.toString(rowNum) + "", ex.getMessage());
        }
    }

    public List<String> getRowValue(int rowNum) {
        if (count() < 0 || count() < rowNum || rowNum <= 0)
            throw exception("Can't Get Row '%s'. [num] > ColumnsCount(%s).", rowNum, count());
        try {
            return select(table.rows().getLineAction(rowNum), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwRowsException(Integer.toString(rowNum), ex.getMessage());
        }
    }

    public final MapArray<String, String> getRowAsText(int rowNum) {
        return getRow(rowNum).toMapArray(IText::getText);
    }

    public final MapArray<String, ICell> getRow(String rowName) {
        try {
            int colsCount = table.columns().count();
            List<WebElement> webRowLine = timer().getResultByCondition(
                    () -> getLineAction(rowName), els -> els.size() == colsCount);
            List<String> headers = table.columns().headers();
            return new MapArray<>(colsCount,
                    table.columns().headers::get,
                    value -> table.cell(webRowLine.get(value), new Column(headers.get(value)), new Row(rowName)));
        } catch (Exception | Error ex) {
            throw throwRowsException(rowName, ex.getMessage());
        }
    }

    protected boolean skipFirstColumn() {
        return hasHeader && lineTemplate == null;
    }

    private RuntimeException throwRowsException(String rowName, String ex) {
        return exception("Can't Get Row '%s'. Reason: %s", rowName, ex);
    }
}